relengapi/blueprints/tooltool/__init__.py

```diff
@@ -0,0 +1,348 @@
+    although the tooltool server will verify the integrity anyway.  The
+    upload must have the header ``Content-Type: application/octet-stream```.
+
+    The query argument ``region=us-west-1`` indicates a preference for URLs
+    in that region, although if the region is not available then URLs in
+    other regions may be returned.
+
+    The returned URLs are only valid for 60 seconds, so all upload requests
+    must begin within that timeframe.  Clients should therefore perform all
+    uploads in parallel, rather than sequentially.  This limitation is in
+    place to prevent malicious modification of files after they have been
+    verified."""
+    region, bucket = get_region_and_bucket(region)
+
+    if not body.message:
+        raise BadRequest("message must be non-empty")
+
+    if not body.files:
+        raise BadRequest("a batch must include at least one file")
+
+    if body.author:
+        raise BadRequest("Author must not be specified for upload")
+    try:
+        body.author = current_user.authenticated_email
+    except AttributeError:
+        raise BadRequest("Could not determine authenticated username")
+
+    # validate permission first
+    visibilities = set(f.visibility for f in body.files.itervalues())
+    for v in visibilities:
+        if not p.get('tooltool.upload.{}'.format(v)).can():
+            raise Forbidden("no permission to upload {} files".format(v))
+
+    session = g.db.session('tooltool')
+    batch = tables.Batch(
+        uploaded=time.now(),
+        author=body.author,
+        message=body.message)
+
+    s3 = current_app.aws.connect_to('s3', region)
+    for filename, info in body.files.iteritems():
+        if info.algorithm != 'sha512':
+            raise BadRequest("'sha512' is the only allowed digest algorithm")
+        if not is_valid_sha512(info.digest):
+            raise BadRequest("Invalid sha512 digest")
+        digest = info.digest
+        file = tables.File.query.filter(tables.File.sha512 == digest).first()
+        if file and file.visibility != info.visibility:
+            raise BadRequest("Cannot change a file's visibility level")
+        if file and file.instances != []:
+            if file.size != info.size:
+                raise BadRequest("Size mismatch for {}".format(filename))
+        else:
+            if not file:
+                file = tables.File(
+                    sha512=digest,
+                    visibility=info.visibility,
+                    size=info.size)
+                session.add(file)
+            log.info("generating signed PUT URL to {} for {}, expires in {}s".format(
+                     info.digest, current_user, UPLOAD_EXPIRES_IN))
+            info.put_url = s3.generate_url(
+                method='PUT', expires_in=UPLOAD_EXPIRES_IN, bucket=bucket,
+                key=util.keyname(info.digest),
+                headers={'Content-Type': 'application/octet-stream'})
```

Comment:
If we need to harden this, we can use Content-MD5 here. See http://docs.aws.amazon.com/AmazonS3/latest/dev/RESTAuthentication.html